src/Fusonic/Linq/OrderedLinq.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+/*
+ * This file is part of Fusonic-linq.
+ *
+ * (c) Fusonic GmbH
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Fusonic\Linq;
+
+use Fusonic\Linq\Helper\LinqHelper;
+use Fusonic\Linq\Iterator\OrderIterator;
+
+/**
+ * Class OrderedLinq
+ * Represents a Linq object that is sorted.
+ * @package Fusonic\Linq
+ */
+class OrderedLinq extends Linq
+{
+    /** @var  OrderIterator */
+    protected $iterator;
+
+    public function __construct($dataSource)
+    {
+        parent::__construct($dataSource);
+    }
+
+    /**
+     * Performs a subsequent ordering of the elements in a sequencing in ascending order according to the key provided by $func.
+     *
+     * @param callback $func    A function to extract a key from an element.
+     * @return OrderedLinq             A new Linq instance whose elements are sorted ascending according to a key.
+     */
+    public function thenBy($func)
+    {
+        $this->iterator->thenBy($func, LinqHelper::LINQ_ORDER_ASC);
```

Comment:
ThenBy must not mutate the original iterator. Instead it should create a new OrderedLinq which uses the original OrderedLinq/iterator as its parent.

Reason: If someone passes you an instance of OrderedLinq, and you mutate it by using thenBy(...), you are changing the callers object and changing his sort behavior.
This is not expected.

It is important that every chained linq method does not modify the state of it's ancestor to avoid unintended side effects.